### PR TITLE
chore(descriptions): children can be component that wraps DescriptionsItem

### LINKED ISSUE #28254: Descriptions描述列表嵌入子组件变成空白的问题
- [ ] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[Edit on CodeSandbox](https://codesandbox.io/s/daibiankuangde-antd492-forked-nse03)

### Steps to reproduce
1.新建一个子组件
2.子组件包含Descriptions.Item
3.Descriptions中引用子组件

### What is expected?
可以显示出来

### What is actually happening?
为空

| Environment | Info |
|---|---|
| antd | 4.9.2 |
| React | React 17.0.1 |
| System | Linux Ubuntu 18.0.4 |
| Browser | Firefox 83.0 (64-bit) |

### PR DESCRIPTION
**Pull request by kerm1it**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

fix https://github.com/ant-design/ant-design/issues/28254

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Children of `Descriptions` allow to be component that wraps `Descriptions.Item`.  |
| 🇨🇳 Chinese | `Descriptions` 的子元素允许是包裹 `Descriptions.Item` 的组件。 |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/descriptions/index.tsx [hunk 1]
```diff
@@ -67,8 +67,30 @@ function getFilledItem(
...
+      } else {
+        childElement = ChildComponent(props);
+      }
+      items.push(...getItems(childElement));
+    }
+  });
+
+  return items;
+}
+
```
Comment (vVvlife) [inline components/descriptions/index.tsx:92]: 赞一个，代码优化得好呀，很多地方我也是写了实现并没有去优化。

**Comment by vVvlife** [comment]
// https://github.com/ant-design/ant-design/issues/28254
  it('Descriptions support function component & class component', () => {
    class TestComponentItem extends React.Component {
      constructor(props) {
        super(props);
        this.state = {};
      }

      render() {
        return (
          <Descriptions.Item label="Test8">Test Data8</Descriptions.Item>
        )
      }
    }
    
    class TestComponent extends React.Component {
        constructor(props) {
            super(props);
            this.state = {};
        }
    
        render() {
            switch (this.props.type) {
            case 'more':
                return(
                <>
                    <Descriptions.Item label="Test6">Test Data6</Descriptions.Item>
                    <Descriptions.Item label="Test7">Test Data7</Descriptions.Item>
                    <TestComponentItem />
                </>
                );
            default:
                return(<Descriptions.Item label="Test5">Test Data5</Descriptions.Item>);
            }
        }
    }
    const TestItem = () => (<Descriptions.Item label="Test4">Test Data4</Descriptions.Item>);
    const Test = ({ type }) => {
      switch(type) {
        case 'more':
          return (
            <>
              <Descriptions.Item label="Test1">Test Data1</Descriptions.Item>
              <Descriptions.Item label="Test2">Test Data2</Descriptions.Item>
              <TestItem />
            </>
          );
        default:
          return(<Descriptions.Item label="Test3">Test Data3</Descriptions.Item>);
      }
    }
    const wrapper = mount(
      <Descriptions title="User Info">
        <Test type="more" />
        <Descriptions.Item label="UserName">Zhou Maomao</Descriptions.Item>
        <Descriptions.Item label="Telephone">1810000000</Descriptions.Item>
        <Descriptions.Item label="Live">Hangzhou, Zhejiang</Descriptions.Item>
        <Descriptions.Item label="Remark">empty</Descriptions.Item>
        <Descriptions.Item label="Address">
          No. 18, Wantang Road, Xihu District, Hangzhou, Zhejiang, China
        </Descriptions.Item>
        <Test />
        <TestComponent type="more" />
        <TestComponent />
      </Descriptions>,
    );
    expect(wrapper.find('td').reduce((total, td) => total + td.props().colSpan, 0)).toBe(13);
    wrapper.unmount();
  })

这个是我之前写的一个用例，你看看合不合适，望指正

File: components/descriptions/index.tsx [hunk 2]
```diff
@@ -67,8 +67,30 @@ function getFilledItem(
   return clone;
 }
 
-function getRows(children: React.ReactNode, column: number) {
-  const childNodes = toArray(children).filter(n => n);
+function getItems(children: React.ReactNode): React.ReactElement[] {
```
Comment (zombieJ) [inline components/descriptions/index.tsx:70]: 语法糖本质上不支持 HOC，内部的元素结构是拆掉用的。遇到这种就要懵逼了：
```jsx
const MyComponent = () => <>
  <Desc.Item />
  <Desc.Item />
  <Desc.Item />
</>;

```

Comment (kerm1it) [reply components/descriptions/index.tsx:70]: 测试里面有写，这种case是可以满足的，会把Item 提出来

**Comment by kerm1it** [comment]
覆盖率怎么没了？

File: components/descriptions/index.tsx (hunk 2 above)
Comment (zombieJ) [reply components/descriptions/index.tsx:70]: thx 看到了，还以为是遍历 `children`，原来是重新渲染。那又会有 context 丢失的问题：
```tsx
const Context = React.createContext(false);

class MyComponent extends React.Component {
  static contextType = Context;

  render() {
    const show = this.context;

    return (
      <>
        {show && <Descriptions.Item label="UserName">Zhou Maomao</Descriptions.Item>}
        <Descriptions.Item label="Telephone">1810000000</Descriptions.Item>
      </>
    );
  }
}

ReactDOM.render(
  <Context.Provider value>
    <Descriptions>
      <MyComponent />
    </Descriptions>
  </Context.Provider>,
  mountNode,
);
```

Comment (zombieJ) [reply components/descriptions/index.tsx:70]: 另一个 case：
```tsx
class MyComponent extends React.Component {
  render() {
    return (
      <>
        <Descriptions.Item label="UserName">Zhou Maomao</Descriptions.Item>
        <Descriptions.Item label="Telephone">1810000000</Descriptions.Item>
      </>
    );
  }
}

const MemoComponent = React.memo(MyComponent, () => true);

ReactDOM.render(
  <Descriptions>
    <MemoComponent />
  </Descriptions>,
  mountNode,
);
```

Comment (vVvlife) [reply components/descriptions/index.tsx:70]: 感觉重新渲染都会带来一些或多或少的问题 T T，只能满足一些基本的业务逻辑

Comment (kerm1it) [reply components/descriptions/index.tsx:70]: 是的，这个就是处理边界情况，如果要满足各种类型组件的渲染，就有点得不偿失了。

Comment (zombieJ) [reply components/descriptions/index.tsx:70]: Context 不能算是边界情况，HOC 只支持一部分反而问答成本会更多。解决语法糖 HOC 问题最好的方式是 render 时的 context 向上注册，只是这需要变更成 mount 后渲染不知道有没有其他副作用。

Comment (kerm1it) [reply components/descriptions/index.tsx:70]: 我意思是，这里就是为了处理子元素不是 Item 的边界情况，这种情况本来就挺少见，如果在这里还要再处理很多 case，那我觉得这个功能就没必要了，直接不支持就行，不然会增加我们的成本，得不偿失。

Comment (vVvlife) [reply components/descriptions/index.tsx:70]: 是的，其实当初我知道有这有问题的时候都是用了renderprops来处理

Comment (zombieJ) [reply components/descriptions/index.tsx:70]: 👌 那还是先把 issue 合到 HOC 大 issue 下关了好了。

**Comment by kerm1it** [comment]
ref: https://github.com/ant-design/ant-design/issues/4853

**Pull request closed without merging**